File: typescipt/src/compiler/JavaScript/builtInFunctions/functionCall.ts
import {ExpressionFunction} from "../../../language/expressions/functions/expressionFunction";
import {FunctionCallExpression} from "../../../language/expressions/functionCallExpression";
import {CodeWriter} from "../writers/codeWriter";

export abstract class FunctionCall {
   public expressionFunction: ExpressionFunction;

   protected constructor(expressionFunction: ExpressionFunction) {
     this.expressionFunction = expressionFunction;
   }

  public static create(expression: FunctionCallExpression): FunctionCall | null {
     return null;
  };


  public renderCustomFunction(codeWriter: CodeWriter) {
  }

  public abstract renderExpression(codeWriter: CodeWriter);
}

File: src/Lexy.Compiler/Compiler/CSharp/BuiltInFunctions/FunctionCall.cs
using Lexy.Compiler.Language.Expressions.Functions;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace Lexy.Compiler.Compiler.CSharp.BuiltInFunctions;

internal abstract class FunctionCall<TFunctionExpression> where TFunctionExpression : FunctionCallExpression
{
    public virtual MemberDeclarationSyntax CustomMethodSyntax(TFunctionExpression expression)
    {
        return null;
    }

    public abstract ExpressionSyntax CallExpressionSyntax(TFunctionExpression expression);
}